app/map.js

```diff
@@ -0,0 +1,205 @@
+var app = require('remote').require('app');
+var jetpack = require('fs-jetpack').cwd(app.getAppPath());
+
+function buildShaderProgram(gl) {
+    var vertexShader = gl.createShader(gl.VERTEX_SHADER);
+    gl.shaderSource(vertexShader, jetpack.read("../app/vertex.glsl"));
+    gl.compileShader(vertexShader);
+    if (!gl.getShaderParameter(vertexShader, gl.COMPILE_STATUS)) {
+        throw new Error("Failed to compile vertex shader: " + gl.getShaderInfoLog(vertexShader));
+    }
+
+    var fragmentShader = gl.createShader(gl.FRAGMENT_SHADER);
+    gl.shaderSource(fragmentShader, jetpack.read("../app/fragment.glsl"));
+    gl.compileShader(fragmentShader);
+    if (!gl.getShaderParameter(fragmentShader, gl.COMPILE_STATUS)) {
+        throw new Error("Failed to compile fragment shader: " + gl.getShaderInfoLog(fragmentShader));
+    }
+
+    var program = gl.createProgram();
+    gl.attachShader(program, vertexShader);
+    gl.attachShader(program, fragmentShader);
+    gl.linkProgram(program);
+    if (!gl.getProgramParameter(program, gl.LINK_STATUS)) {
+        throw new Error("Failed to link shader program: " + gl.getProgramInfoLog(program));
+    }
+
+    return program;
+}
+
+function buildTileDataTexture(data) {
+    var out = new Uint8Array(data.length * 4);
+    for (var i = 0; i < data.length; i++) {
+        var t = data[i];
+        out[i * 4 + 0] = t % 256;
+        out[i * 4 + 1] = (t >> 8) % 256;
+        out[i * 4 + 2] = (t >> 16) % 256;
+        out[i * 4 + 3] = (t >> 24) % 256;
+    }
+    return out;
+}
+
+export var Map = function(mapfile, mapdatafile, vspfile) {
+    console.log("Loading map", mapfile);
+
+    this.readyPromise = new Promise(function(resolve, reject) {
+        this.promiseResolver = resolve;
+        this.promiseRejecter = reject;
+    }.bind(this));
+
+    this.mapPath = mapfile;
+    this.mapData = jetpack.read(mapfile, 'json');
+    this.renderString = this.mapData.renderstring.split(",");
+    this.mapSize = [0,0];
+    for (var i = 0; i < this.mapData.layers.length; i++) {
+        if (this.mapData.layers[i].dimensions.X > this.mapSize[0]) this.mapSize[0] = this.mapData.layers[i].dimensions.X;
+        if (this.mapData.layers[i].dimensions.Y > this.mapSize[1]) this.mapSize[1] = this.mapData.layers[i].dimensions.Y;
+    }
+    this.camera = [0, 0, 1];

```

Comment:
Not quite -- this.camera is [x, y, scale factor]. The x and y coordinates are pixel-based, and the scale factor is just a float from 0.125 to 16 (in the current zoom handling anyway).